Python/cheritrace.i

```diff
@@ -1,80 +1,250 @@
 // swig interface definition for the cheritrace library
 
-%module(directors="1") pycheritrace
+%module pycheritrace
 
 %include "std_string.i";
 %include "std_array.i"
 %include "std_shared_ptr.i"
 %include "stdint.i"
+%include "exception.i"
 
 %include "std_bitset.i"
 
 %{
 #define SWIG_FILE_WITH_INIT
 #include "../disassembler.hh"
-#include "callback.hh"
+#include "../streamtrace.hh"
 #include <functional>
 %}
 
 // unsupported by SWIG
 #define __attribute__(x)
 #define static_assert(x,y)
 
+/* XXX Ignore trace::(string, notifier) as the notifier callback need
+ *  some more work.
+ * The notifier is saved in the thread, we must Py_INCREF the object!
+ * may need to acquire the GIL from the loading thread
+ */
+%ignore cheri::streamtrace::trace::open;
+%rename("%s") cheri::streamtrace::trace::open(const std::string&);
+
 // wrap STL types
 %shared_ptr(cheri::streamtrace::trace)
 %shared_ptr(cheri::streamtrace::trace_view)
 %template (GprArray) std::array<unsigned long long, 31>;
 %template (CapArray) std::array<struct cheri::streamtrace::capability_register, 32>;
 %template (GprBitset) std::bitset<31>;
 %template (CapBitset) std::bitset<32>;
+/* Workaround missing support of nested unions (see accessors below) */
+%ignore cheri::streamtrace::debug_trace_entry::reg_value;
+
+/* 
+ * Convert from python function to C function pointer for trace::scanner
+ * trace::detailed_scanner, trace::filter and trace::notifier
+ */
+%typemap (in)
+cheri::streamtrace::trace::scanner
+{
+     if (!PyCallable_Check($input)) {
+	  SWIG_exception(SWIG_TypeError, "Object not callable");
+     }
+     $1 = [$input](cheri::streamtrace::debug_trace_entry entry, uint64_t idx) {
```

Comment:
It is a PyObject*, it points to the python callback given as the first argument to the scan function. It is the same in all the other typemaps.